Lapack geqrf

lucbv opened the pull request:
Taking over PR #2205 rebasing on the current state of develop and re-running tests, will also have a look at the unit-tests and might add more checks.

eeprude committed:
Lapack - geqrf: adding wrapper for QR decomposition

The wrappers will allow use of LAPACK, cusolver, rocsolver and
magma

Signed-off-by: Luc-Berger Vergiat <lberge@sandia.gov>

jgfouca commented on lapack/src/KokkosLapack_geqrf.hpp:
```diff
@@ -0,0 +1,152 @@
+  static_assert(std::is_same_v<typename InfoArray::non_const_value_type, int>,
+                "KokkosLapack::geqrf: Info must be an array of integers.");
+
+  int64_t m     = A.extent(0);
```
These could all be consts.

jgfouca commented on lapack/unit_test/Test_Lapack_geqrf.hpp:
```diff
@@ -0,0 +1,602 @@
+    for (int index(k + 1); index < m; ++index) {
+      h_v[index] = h_A(index, k);
+    }
+#if 0  // def HAVE_KOKKOSKERNELS_DEBUG
```
Is it worthwhile to keep these `#if 0` blocks around? Should we switch these to something like `#if QEQRF_DEBUG` so that the programmer can turn these all back on easily?

jgfouca commented on lapack/tpls/KokkosLapack_Host_tpl.cpp:
```diff
@@ -51,6 +51,17 @@ void F77_BLAS_MANGLE(strtri, STRTRI)(const char*, const char*, int*, const float
+/// Geqrf
+///
+
+void F77_BLAS_MANGLE(sgeqrf, SGEQRF)(int*, int*, float*, int*, float*, float*, int*, int*);
```
I noticed none of these are consts. Is every argument an output or inout?

jgfouca reviewed (changes requested):
Looks good. Just a few minor things.

lucbv committed:
Lapack - QR: adding documentation

Signed-off-by: Luc Berger-Vergiat <lberge@sandia.gov>

lucbv committed:
Lapack - QR: fixing documentation

Signed-off-by: Luc Berger-Vergiat <lberge@sandia.gov>

lucbv commented:
@jgfouca @vqd8a this is ready for review now : )

jgfouca commented:
@lucbv , I've had some good results with co-pilot reviews on my other projects, so I asked for one here.

lucbv commented:
@jgfouca sure, that's not a bad idea and we anyway have to wait for CI to get back online before moving ahead with this so we have time to fix the issues it detected.

lucbv committed:
Lapack - geqrf: resolving co-pilot review

Signed-off-by: Luc Berger-Vergiat <lberge@sandia.gov>

lucbv committed:
Lapack - geqrf: applying clang-format

Signed-off-by: Luc Berger-Vergiat <lberge@sandia.gov>

lucbv commented:
Okay, co-pilot pointed some interesting things, I think I got it all squared away.

lucbv committed:
Lapack - QR: fixing HostMirror to host_mirror_type

Also simplifying some memory space specification, there might still
be an issue with a View constructor though...

Signed-off-by: Luc Berger-Vergiat <lberge@sandia.gov>

lucbv committed:
Lapack - QR: adding support for Accelerate

Adding support for Apple library and fixing logic error leading
to segfault on Host.

Signed-off-by: Luc Berger-Vergiat <lberge@sandia.gov>

lucbv committed:
Lapack - QR: applying clang-format

Signed-off-by: Luc Berger-Vergiat <lberge@sandia.gov>
